ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/util/DebugDataDumper.java

```diff
@@ -0,0 +1,157 @@
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+
+public class DebugDataDumper {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private static final String GOSSIP_MESSAGES_DIR = "gossip_messages";
+  private static final String DECODING_ERROR_SUB_DIR = "decoding_error";
+  private static final String REJECTED_SUB_DIR = "rejected";
+  private static final String INVALID_BLOCK_DIR = "invalid_blocks";
+
+  private boolean enabled;
+  private final Path directory;
+
+  public DebugDataDumper(final Path directory, final boolean enabled) {
+    this.enabled = enabled;
+    this.directory = directory;
+    if (!enabled) {
+      return;
+    }
+
+    final Path gossipMessagesPath = this.directory.resolve(GOSSIP_MESSAGES_DIR);
+    createDirectory(gossipMessagesPath, GOSSIP_MESSAGES_DIR, "gossip messages");
+    createDirectory(
+        gossipMessagesPath.resolve(DECODING_ERROR_SUB_DIR),
+        DECODING_ERROR_SUB_DIR,
+        "gossip messages with decoding errors");
+    createDirectory(
+        gossipMessagesPath.resolve(REJECTED_SUB_DIR), REJECTED_SUB_DIR, "rejected gossip messages");
+    createDirectory(this.directory.resolve(INVALID_BLOCK_DIR), INVALID_BLOCK_DIR, "invalid blocks");
+  }
+
+  public void saveGossipMessageDecodingError(
+      final String topic, final Optional<UInt64> arrivalTimestamp, final Bytes originalMessage) {
+    if (!enabled) {
+      return;
+    }
+    final String fileName = String.format("%s.ssz", formatTimestamp(arrivalTimestamp));
+    final Path topicPath =
+        Path.of(GOSSIP_MESSAGES_DIR).resolve(DECODING_ERROR_SUB_DIR).resolve(topic);
+    saveBytesToFile(
+        "gossip message with decoding error", topicPath.resolve(fileName), originalMessage);
+  }
+
+  public void saveGossipRejectedMessageToFile(
+      final String topic, final Optional<UInt64> arrivalTimestamp, final Bytes decodedMessage) {
+    if (!enabled) {
+      return;
+    }
+    final String fileName = String.format("%s.ssz", formatTimestamp(arrivalTimestamp));
+    final Path topicPath = Path.of(GOSSIP_MESSAGES_DIR).resolve(REJECTED_SUB_DIR).resolve(topic);
+    saveBytesToFile("rejected gossip message", topicPath.resolve(fileName), decodedMessage);
+  }
+
+  public void saveInvalidBlockToFile(
+      final UInt64 slot, final Bytes32 blockRoot, final Bytes blockSsz) {
+    if (!enabled) {
+      return;
+    }
+    final String fileName = String.format("%s_%s.ssz", slot, blockRoot.toUnprefixedHexString());
+    saveBytesToFile("invalid block", Path.of(INVALID_BLOCK_DIR).resolve(fileName), blockSsz);
+  }
+
+  @VisibleForTesting
+  protected void saveBytesToFile(
+      final String object, final Path relativeFilePath, final Bytes bytes) {
+    final Path path = directory.resolve(relativeFilePath);
+    try {
+      Files.write(path, bytes.toArray());
+      LOG.info("Saved {} bytes to file. Location: {}", object, relativeFilePath);
```

Comment:
we're going to be constantly getting
```
Saved rejected gossip message to file. location: invalid_blocks/1234_badbeefbadbeef.ssz
Saved rejected gossip message to file. location: invalid_blocks/5432_aabbccddeeffaa.ssz
```
and by the thousands potentially for gossip topics. this is going to be super noisy and potentially less useful.
I do agree that its potentially the 1 place we want info rather than debug, or at the level above where we can log more informatively...

For blocks i'd log something to the effect of 
`Saved invalid block at slot 12345(badbeefbadbeef)`
we don't really need the filename because its well defined and a large component of noise.

I'd probably suggest similar for gossip, logging.
`Saved rejected gossip message on topic {} at {}` 
last field being the formatted time or something...

The reason this is important is we need to be able to refer to these log messages for source data, and we dont want them more noisy than htey need to be.